docs: guide for dependency imports

BenjaminMummery opened the pull request:
[ORQSDK-487]

# The problem

Imports to tasks can be complicated. The addition of PythonImports may make this more confusing for our users.

# This PR's solution

A user facing docs page detailing the best practices for workflow dependencies.

# Checklist

_Check that this PR satisfies the following items:_

- [x] Tests have been added for new features/changed behavior (if no new features have been added, check the box).
- [x] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [x] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`


[ORQSDK-487]: https://zapatacomputing.atlassian.net/browse/ORQSDK-487?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

BenjaminMummery committed:
docs: guide for dependency imports

[ORQSDK-487]

BenjaminMummery committed:
docs: update changelog with depency installation guide

[ORQSDK-487]

jamesclark-Zapata commented on docs/guides/dependency-installation.rst:
```diff
@@ -0,0 +1,223 @@
+The local importer is the default source importer where no other importer is specified. It imports the current repo as a module, making its methods available to tasks in local workflow runs.
+
+
+Good Practice
+-------------
+
+For workflows running remotely the ``GitHubImport`` importer should be used unless there is a specific reason to do otherwise (i.e. the code is hosted somewhere other than GitHub). ``GitImport.infer()`` is highly user-specific and so should never be used in a collaborative context.
+
+For workflows running locally, the task decorator can typically be used without specifying the source importer as the SDK will set this appropriately for the context:
+
+.. code-block:: python
+    @sdk.task
+    def my_task():
+        pass
+
+will perform a local import unless running in a Jupyter Notebook (or other interactive session) in which case am inline import will be performed.
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** ‏Should this be first?

BenjaminMummery replied:
I wAlmost certainly, especially with the quick reference tl;dr up top. Fixed in [2337332](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74/commits/2337332f42a2571eead93827cf1d58ae61ec6a0e)

jamesclark-Zapata reviewed (changes requested):
woohoo! Thanks for tackling these docs. This will certainly help our users.

A few minor comments + I'm not sure on the word "ingests" in some places.

Can you ask Tim or some internal users (from the recent Slack thread) to check this too?

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
docs: move Good Practice section to beginning of guide.

[ORQSDK-487]
Suggested by https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74#discussion_r1131075669

BenjaminMummery committed:
docs: correct Runtime column for inline import.

fixes https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74#discussion_r1131044425

[ORQSDK-487]

BenjaminMummery committed:
docs: remove reference to local clone of repo.

[ORQSDK-487]
Fixes https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74#discussion_r1131043393

BenjaminMummery commented on docs/guides/dependency-installation.rst:
```diff
@@ -0,0 +1,223 @@
+      - ``GitHubImport``
+      - remote
+      - On Ray runtimes, secrets can be used to access private repos.
+    * - Non-github git repo without a local clone
```
This made more sense when the table included `.infer`, I should have removed it. Fixed in
[f419250](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74/commits/f41925080adb712979d7961a0c5eff4822c9aef6)

BenjaminMummery commented on docs/guides/dependency-installation.rst:
```diff
@@ -0,0 +1,223 @@
+      -
+    * - Jupyter Notebook
+      - ``InlineImport``
+      - local
```
Fixed in [d594f39](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74/commits/d594f39c00cbbd912d31fa490713bdc0c5339fb6)

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-487-a-guide-dependency-installation-sources-order-good-practices

BenjaminMummery committed:
docs: update quick reference table to reflect the fact that PyPI imports are downloaded locally.

[ORQSDK-487]

BenjaminMummery commented on docs/guides/dependency-installation.rst:
```diff
@@ -0,0 +1,223 @@
+      -
+    * - PyPi
+      - ``PythonImports``
+      - remote
```
Updated in [4502ba6](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74/commits/4502ba653b15e3828263d6f7c35068b9eeaefffc) so this is now correct w.r.t. the current behaviour.

BenjaminMummery committed:
docs: remove all instances of the word 'ingest'

[ORQSDK-487]

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
Update docs/guides/dependency-installation.rst

Co-authored-by: James Clark <70290797+jamesclark-Zapata@users.noreply.github.com>

BenjaminMummery committed:
docs: address https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74#discussion_r1131052373

[ORQSDK-487]

BenjaminMummery commented on docs/guides/dependency-installation.rst:
```diff
@@ -0,0 +1,223 @@
+    def demo_task():
+        pass
+
+The PAT is ingested from the portal secrets management facility at run time, and installation of the import proceeds as usual.
```
[2b0e649](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/74/commits/2b0e6498e022a5ff3b92ab831d69af6fe78db3c1)

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-487-a-guide-dependency-installation-sources-order-good-practices
